manuals/fvdemount.1
.Dd July  5, 2026
.Dt FVDEMOUNT 1
.Os
.Sh NAME
.Nm fvdemount
.Nd mounts a Core Storage (CS) volume
.Sh SYNOPSIS
.Nm fvdemount
.Op Fl e Ar plist_path
.Op Fl k Ar key
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl r Ar recovery_password
.Op Fl X Ar extended_options
.Op Fl huvV
.Ar sources
.Ar mount_point
.Sh DESCRIPTION
.Nm fvdemount
is a utility to mount a Core Storage (CS) volume
.Pp
.Nm fvdemount
is part of the
.Nm libfvde
package.
.Nm libfvde
is a library to access the Core Storage (CS) format
.Pp
.Ar sources
one or more source files or devices.
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl e Ar plist_path
specify the path of the EncryptedRoot.plist.wipekey file
.It Fl h
shows this help
.It Fl k Ar key
specify the volume master key formatted in base16
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl r Ar recovery_password
specify the recovery password (or passphrase)
.It Fl u
unattended mode (disables user interaction)
.It Fl v
verbose output to stderr, while fvdemount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# fvdemount image1.raw /mnt/fvde
fvdemount 20260705
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr fvdeinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libfvde/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* format support
  - handle volume offset detection of partial encrypted logical volume
  - use the VMK signature (a6 a6 a6 ...) to indicate the volume is unlocked not the unencrypted volume header?
  - volume with no 0x001a, 0x0305 and 0x0505 ?
  - remove libfvde_logical_volume_descriptor_get_last_block_number after debugging unsupported format variants?

* libfvde
  - libfvde_segment_descriptor_compare check for overlapping ranges
  - clean up: split off AES KEK functions
  - add handling of FVDE volumes with corrupt metadata, e.g. file header
  - store passwords as hashes?
  - read and compare secondary (physical) volume header
  - support out-of-order physical volume file IO pool

* add debug output
  - encryption method
  - checksum algorithm

* pyfvde
  - add functionality to set encrypted context plist from a file-like object

* fvdeinfo
  - print logical volume name and/or identifier when asking for a password

* fvdemount
  - print logical volume name and/or identifier when asking for a password

* code clean up
  - remove deprecated functions from libfvde and pyfvde
  - remove libfvde_volume_get_logical_volume_encryption_method
  - rename FileVault to CoreStorage ?
  - check for leading spaces
  - check for () without spaces
  - check for = alignment
  - check for if ( instead of if(
  - check error codes

* check README files
  - make sure pkg-config is mentioned as a dependency

20141018
* see `git log' for more recent change log
* removed README.macosx

20141017
* changes for deployment

20140907
* updated dependencies
* updated msvscpp files, not operational yet

20140531
* updated dependencies

20140112
* 2014 update
* updated dependencies
* exposed some encryption context plist functions in API

20130909
* updated dependencies
* worked on libcthreads build support

20131205
* updated dependencies

20130716
* updated dependencies
* replaced AES-XTS code by libcaes XTS implementation
* worked on git support
* code clean up

20130703
* code clean up

20130412
* updated README files
* updated dependencies
* added tests directory

20130331
* updated dependencies
* code clean up

20130305
* added safe-guards for XML plist data
* added check for block serial number
* added support for 0x0505 block that is defined before 0x001a block

20130228
* worked on improving logical volume offset detection

20130227
* worked on improving logical volume size detection
* code clean up

20130226
* worked on improving logical volume size detection
* code clean up

20130221
* fix for value overwrite of logical volume size by successive block

20130220
* small change to include/types.h

20130113
* 2013 update
* updated dependencies
* updated msvscpp files

20121217
* bug fixes

20121212
* updates to debug output

20121114
* updated libfvde.3 man page
* added volume get offset function
* worked on set keys
* code clean up
* worked on msvscpp files

20121113
* updated man pages
* added change to support number of iterations in password
  wrapped kek

20121112
* updated dependencies
* replaced libfvde_parse_uuid_string by libfguid function
* fixes in pc and spec file

20120916
* worked on fvdemount

20120915
* applied updates
* worked on fvdemount
  - added -X option
  - replaced EBADFD by EIO

20120813
* worked on encrypted removable media
* worked on man pages
* fixed issue for debug output for libfvde_io_handle_read_unencrypted_volume_header
* worked on function to pass VMK directly

20120808
* applied updates

20120724
* code clean up

20120723
* code clean up
* removed key and volume master key files they served no purpose in contrast
  to original libbde code

20120722
* code clean up
* fixed potential memory and data leak in set password functions

20120721
* fix for acinclude.m4 and libxml2.m4
* added libcsplit
* applied updates
* switched library order in tools
* updated .pc and .spec file
* preparations for msvscpp build

20120521
* configure.ac fixes for dpkg files

20120520
* code clean up

20120519
* removed libfvalue
* code clean up
  - renamed plist
  - change naming of volume offset and size
* add support for secondary encrypted metadata block
* added XML parsing to encrypted metadata block

20120518
* code clean up of fvdemount
* code clean up
  - removed more unused code
  - fix for relative volume offset

20120516
* bug fixes for plist
* code clean up
* added more debug output

20120514
* fixes for wide character build

20120510
* updated dependencies
* code clean up

20120405
* fix for libxml2 detection on MacOS-X

20120208
* code clean up

20120207
* code clean up
* fixed wrong context for decryption

20120206
* code clean up

20120202
* code clean up

20120124
* added debug output
* rewrite of dealing with metadata
* removed AES CCM code
* removed BDE metadata definitions

20120117
* added debug output

20120116
* 2012 update
* updated pc file
* updated dependencies
* removed libbde
* renamed to libfvde
* added libfcache
* updates for array type changes
* updates for libcaes changes
* updates for libbfio changes
* updated for libfcache/libfdata changes
* fixed a several potential memory leaks

fvdetools/fvdemount.c
/*
 * Mounts a Core Storage (CS) volume.
 *
 * Copyright (C) 2011-2026, Omar Choudary <choudary.omar@gmail.com>,
 *                          Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "fvdetools_getopt.h"
#include "fvdetools_i18n.h"
#include "fvdetools_libcerror.h"
#include "fvdetools_libclocale.h"
#include "fvdetools_libcnotify.h"
#include "fvdetools_libfvde.h"
#include "fvdetools_output.h"
#include "fvdetools_signal.h"
#include "fvdetools_unused.h"
#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"

mount_handle_t *fvdemount_mount_handle = NULL;
int fvdemount_abort                    = 0;

/* Signal handler for fvdemount
 */
void fvdemount_signal_handler(
      fvdetools_signal_t signal FVDETOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "fvdemount_signal_handler";

	FVDETOOLS_UNREFERENCED_PARAMETER( signal )

	fvdemount_abort = 1;

	if( fvdemount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     fvdemount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use fvdemount to mount a Core Storage (CS) volume.";

	fvdetools_option_t options[ ] = {
		{ 'e', "plist_path", "specify the path of the EncryptedRoot.plist.wipekey file" },
		{ 'h', NULL, "shows this help" },
		{ 'k', "key", "specify the volume master key formatted in base16" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'r', "recovery_password", "specify the recovery password (or passphrase)" },
		{ 'u', NULL, "unattended mode (disables user interaction)" },
		{ 'v', NULL, "verbose output to stderr, while fvdemount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a FileVault Drive Encrypted (FVDE) volume" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	system_character_t * const *sources          = NULL;
	const system_character_t *path_prefix        = NULL;
	libfvde_error_t *error                       = NULL;
	size_t path_prefix_size                      = 0;
	system_character_t *option_key               = NULL;
	system_character_t *option_offset            = NULL;
	system_character_t *option_password          = NULL;
	system_character_t *option_plist_path        = NULL;
	system_character_t *option_recovery_password = NULL;
	char *program                                = "fvdemount";
	system_integer_t option                      = 0;
	int number_of_options                        = (int) ( sizeof( options ) / sizeof( fvdetools_option_t ) );
	int number_of_sources                        = 0;
	int unattended_mode                          = 0;
	int verbose                                  = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point              = NULL;
	int result                                   = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations fvdemount_fuse_operations;

	system_character_t *option_extended_options  = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                         = { program, NULL };
	struct fuse_args fvdemount_fuse_arguments    = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args fvdemount_fuse_arguments    = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *fvdemount_fuse_channel     = NULL;
#endif
	struct fuse *fvdemount_fuse_handle           = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS fvdemount_dokan_operations;
	DOKAN_OPTIONS fvdemount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "fvdetools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( fvdetools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	fvdetools_output_version_fprint(
	 stdout,
	 program );

	if( fvdetools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = fvdetools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				fvdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'e':
				option_plist_path = optarg;

				break;

			case (system_integer_t) 'h':
				fvdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'k':
				option_key = optarg;

				break;

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'r':
				option_recovery_password = optarg;

				break;

			case (system_integer_t) 'u':
				unattended_mode = 1;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				fvdetools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source files or devices.\n" );

		fvdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	sources           = &( argv[ optind ] );
	number_of_sources = argc - optind - 1;

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		fvdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ argc - 1 ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libfvde_notify_set_stream(
	 stderr,
	 NULL );
	libfvde_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &fvdemount_mount_handle,
	     unattended_mode,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_plist_path != NULL )
	{
		if( mount_handle_set_encrypted_root_plist(
		     fvdemount_mount_handle,
		     option_plist_path,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set encrypted root plist.\n" );

			goto on_error;
		}
	}
	if( option_key != NULL )
	{
		if( mount_handle_set_key(
		     fvdemount_mount_handle,
		     option_key,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set key.\n" );

			goto on_error;
		}
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     fvdemount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( mount_handle_set_password(
		     fvdemount_mount_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
	if( option_recovery_password != NULL )
	{
		if( mount_handle_set_recovery_password(
		     fvdemount_mount_handle,
		     option_recovery_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set recovery password.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\FVDE" );
#else
	path_prefix = _SYSTEM_STRING( "/fvde" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     fvdemount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( mount_handle_open(
	     fvdemount_mount_handle,
	     sources,
	     number_of_sources,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
/* TODO add is locked support
	if( mount_handle_is_locked(
	     fvdemount_mount_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to unlock source volume\n" );

		goto on_error;
	}
*/
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		fvdemount_fuse_arguments.argc = 0;
		fvdemount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &fvdemount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fvdemount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &fvdemount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &fvdemount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	fvdemount_fuse_operations.open       = &mount_fuse_open;
	fvdemount_fuse_operations.read       = &mount_fuse_read;
	fvdemount_fuse_operations.release    = &mount_fuse_release;
	fvdemount_fuse_operations.opendir    = &mount_fuse_opendir;
	fvdemount_fuse_operations.readdir    = &mount_fuse_readdir;
	fvdemount_fuse_operations.releasedir = &mount_fuse_releasedir;
	fvdemount_fuse_operations.getattr    = &mount_fuse_getattr;
	fvdemount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	fvdemount_fuse_handle = fuse_new(
	                         &fvdemount_fuse_arguments,
	                         &fvdemount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         fvdemount_mount_handle );

	if( fvdemount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          fvdemount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	fvdemount_fuse_channel = fuse_mount(
	                          mount_point,
	                          &fvdemount_fuse_arguments );

	if( fvdemount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	fvdemount_fuse_handle = fuse_new(
	                         fvdemount_fuse_channel,
	                         &fvdemount_fuse_arguments,
	                         &fvdemount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         fvdemount_mount_handle );

	if( fvdemount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          fvdemount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 fvdemount_fuse_handle );

	fuse_destroy(
	 fvdemount_fuse_handle );

	fuse_opt_free_args(
	 &fvdemount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &fvdemount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &fvdemount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	fvdemount_dokan_options.Version    = DOKAN_VERSION;
	fvdemount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	fvdemount_dokan_options.SingleThread = TRUE;
#else
	fvdemount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		fvdemount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		fvdemount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	fvdemount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	fvdemount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	fvdemount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	fvdemount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	fvdemount_dokan_operations.CreateDirectory      = NULL;
	fvdemount_dokan_operations.Cleanup              = NULL;
	fvdemount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fvdemount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fvdemount_dokan_operations.WriteFile            = NULL;
	fvdemount_dokan_operations.FlushFileBuffers     = NULL;
	fvdemount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fvdemount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fvdemount_dokan_operations.FindFilesWithPattern = NULL;
	fvdemount_dokan_operations.SetFileAttributes    = NULL;
	fvdemount_dokan_operations.SetFileTime          = NULL;
	fvdemount_dokan_operations.DeleteFile           = NULL;
	fvdemount_dokan_operations.DeleteDirectory      = NULL;
	fvdemount_dokan_operations.MoveFile             = NULL;
	fvdemount_dokan_operations.SetEndOfFile         = NULL;
	fvdemount_dokan_operations.SetAllocationSize    = NULL;
	fvdemount_dokan_operations.LockFile             = NULL;
	fvdemount_dokan_operations.UnlockFile           = NULL;
	fvdemount_dokan_operations.GetFileSecurity      = NULL;
	fvdemount_dokan_operations.SetFileSecurity      = NULL;
	fvdemount_dokan_operations.GetDiskFreeSpace     = NULL;
	fvdemount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fvdemount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	fvdemount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	fvdemount_dokan_operations.Cleanup              = NULL;
	fvdemount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	fvdemount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	fvdemount_dokan_operations.WriteFile            = NULL;
	fvdemount_dokan_operations.FlushFileBuffers     = NULL;
	fvdemount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	fvdemount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	fvdemount_dokan_operations.FindFilesWithPattern = NULL;
	fvdemount_dokan_operations.SetFileAttributes    = NULL;
	fvdemount_dokan_operations.SetFileTime          = NULL;
	fvdemount_dokan_operations.DeleteFile           = NULL;
	fvdemount_dokan_operations.DeleteDirectory      = NULL;
	fvdemount_dokan_operations.MoveFile             = NULL;
	fvdemount_dokan_operations.SetEndOfFile         = NULL;
	fvdemount_dokan_operations.SetAllocationSize    = NULL;
	fvdemount_dokan_operations.LockFile             = NULL;
	fvdemount_dokan_operations.UnlockFile           = NULL;
	fvdemount_dokan_operations.GetFileSecurity      = NULL;
	fvdemount_dokan_operations.SetFileSecurity      = NULL;
	fvdemount_dokan_operations.GetDiskFreeSpace     = NULL;
	fvdemount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	fvdemount_dokan_operations.Unmounted            = NULL;
	fvdemount_dokan_operations.FindStreams          = NULL;
	fvdemount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &fvdemount_dokan_options,
	          &fvdemount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Core Storage (CS) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( fvdemount_fuse_handle != NULL )
	{
		fuse_destroy(
		 fvdemount_fuse_handle );
	}
	fuse_opt_free_args(
	 &fvdemount_fuse_arguments );
#endif
	if( fvdemount_mount_handle != NULL )
	{
		mount_handle_free(
		 &fvdemount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

